Sprint 2/step 2

maybe777 opened the pull request:
Попытка запрыгнуть в последний вагон уходящего поезда) Будьте снисходительны)))

maybe-nl committed:
sprint-2/step-1

1) BurgerConstructor with has GlobalContext now

2) Order number from API

3) Order sum handled automatically

maybe777 committed:
sprint-2/step-1

1) BurgerConstructor with has GlobalContext now

2) Order number from API

3) Order sum handled automatically

gennady-bars commented on src/redux/reducers/order-reducer.tsx:
```diff
@@ -0,0 +1,36 @@
+
+const orderReducer = (state: IOrderState = initialState, action: any) => {
+    switch (action.type) {
+        case "GET_ORDER_NUMBER":
```
`"GET_ORDER_NUMBER"`


Каждый тип экшена должен быть вынесен в константу в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const GET_ORDER_NUMBER = 'GET_ORDER_NUMBER'`


**Нужно исправить это во всем коде проекта**

maybe777 replied:
Поправил

gennady-bars commented on src/redux/actions/order-actions.tsx:
```diff
@@ -0,0 +1,38 @@
+                    });
+                } else {
+                    dispatch({
+                        type: GET_ORDER_ERROR
```
`                        type: GET_ORDER_ERROR
`

**Можно лучше**

* Чтобы не дублировать типы экшенов и сами объекты с `type`, принято выносить их в отдельные функции, называемые `action creator` (функцию, которая возвращает объект с `type`). Это позволяет сделать понятное название для функции, которое будет говорить, что делает она, и избавит от повтора объекта с `type` в коде. Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)

maybe777 replied:
Поправил, спасибо.

gennady-bars commented on src/api/api.tsx:
```diff
@@ -0,0 +1,53 @@
+import React from "react";
+
+
+const ORDER_URL: string = "https://norma.nomoreparties.space/api/orders"
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.

maybe777 replied:
Поправил

gennady-bars commented on src/api/api.tsx:
```diff
@@ -0,0 +1,53 @@
+    let result: BurgerItem[] = [];
+
+    await fetch(INGREDIENTS_URL)
+        .then(res => res.json())
```
Любой ответ от сервера нужно проверять на `ok` и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.


```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется этот код во всех запросах. Чтобы этого избежать, нужно сделать отдельный метод `checkResponse`, который будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

и теперь можно удалить кучу ненужно повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только `ссылка` на метод. Не нужно его вызывать. Он сам вызовется, так как в `then` нужно передавать именно `функцию`, а не `вызов` функции.

maybe777 replied:
Крутая практика! Спасибо, поправил.

gennady-bars commented on src/api/api.tsx:
```diff
@@ -0,0 +1,53 @@
+                throw new Error("Loading error.")
+            }
+        })
+        .catch(e => {
```
Лучше не использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится в `thunk`-функциях после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.

maybe777 replied:
Да, так и есть, поправил.

gennady-bars commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -28,53 +83,39 @@ export default function BurgerConstructor(props: IBurgerConstructor) {
                     <ul className={styles.list}>
-                        {props.data.map((item, index) => (
+                        {INGREDIENTS.map((item: BurgerItem | any, index: number) => (
                             <li key={'novice_item_' + index} className={styles.item}>
```
`key={'novice_item_' + index}`

для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время сброса в зону `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid, Или Вам должны были показать, как это сделать, иначе некорректно может работать DND. 

А в `key` будет уже просто **key={item.uuid}**


--------

[Почему индекс не нужно использовать, можно прочитать тут:]( https://medium.com/russian/%D0%B8%D0%BD%D0%B4%D0%B5%D0%BA%D1%81-%D0%B2-%D0%BA%D0%B0%D1%87%D0%B5%D1%81%D1%82%D0%B2%D0%B5-%D0%BA%D0%BB%D1%8E%D1%87%D0%B0-%D0%B0%D0%BD%D1%82%D0%B8-%D0%BF%D0%B0%D1%82%D1%82%D0%B5%D1%80%D0%BD-56bab8ada309)

maybe777 replied:
Поправил

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)


К сожалению, работу придется отклонить без потери итерации, так как проект не запускается. Скрин ошибки http://joxi.ru/E2pZKygHvYXdyr

Еще успел заметить ошибки:

* Каждый тип экшена должен быть вынесен в константу в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const GET_ORDER_NUMBER = 'GET_ORDER_NUMBER'`
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.
* Любой ответ от сервера нужно проверять на `ok` и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.
* нужно сделать отдельный метод `checkResponse`, который будет проверять ответ от сервера
* Лучше не использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится в `thunk`-функциях после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.
* для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время сброса в зону `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid, Или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={item.uuid}**






**Можно лучше**


* Чтобы не дублировать типы экшенов и сами объекты с `type`, принято выносить их в отдельные функции, называемые `action creator` (функцию, которая возвращает объект с `type`). Это позволяет сделать понятное название для функции, которое будет говорить, что делает она, и избавит от повтора объекта с `type` в коде. Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)

maybe777 committed:
Поправил замечание первой итерации.

- Добавлен uuid для списка в constructor.tsc

- исправлено* использование переменных экшенов в редьюсерах

- все дублирующие участки кода заменены на функции

gennady-bars commented on src/redux/actions/order-actions.tsx:
```diff
@@ -0,0 +1,38 @@
+                        payload: res
+                    });
+                } else {
+                    dispatch(orderError);
```
`                    dispatch(orderError);
`

чтобы экшен-криейтер создал экшен, нужно его вызывать


```js


 dispatch(orderError());
```

gennady-bars commented on src/redux/actions/order-actions.tsx:
```diff
@@ -0,0 +1,38 @@
+            })
+            .then(() => console.log("Get Order number success."))
+            .catch(err => {
+                dispatch(orderError);
```
`                    dispatch(orderError);
`

чтобы экшен-криейтер создал экшен, нужно его вызывать


```js


 dispatch(orderError());
```

gennady-bars commented on src/redux/actions/details-actions.tsx:
```diff
@@ -0,0 +1,23 @@
+}
+
+export function closeDetails(item: BurgerItem) {
+    return function (dispatch: any) {
```
**Можно лучше**

`thunk`-функции создаются обычно для асинхронных запросов на сервер, а для обычных экшенов делают обычные `action creator` (функцию, которая возвращает объект с `type`). И главный показатель – это то, что если внутри используется только 1 раз `dispatch`, то уже можно сделать  `action creator`, а не `thunk`


Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)


Вот так тут будет в виде экшен-криейтера (в коде ничего менять не нужно нигде больше)

```js

export function closeDetails(item: BurgerItem) {
    return {
        type: CLOSE_DETAILS,
        item: item
    }
}


```

gennady-bars commented on src/api/api.tsx:
```diff
@@ -0,0 +1,52 @@
+    return result;
+}
+
+export function checkResponse(res: any){
```
`any`


для типизации ответа от сервера есть встроенный тип `Response`


```js

export function checkResponse(res: Response){


```

gennady-bars commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -28,53 +81,39 @@ export default function BurgerConstructor(props: IBurgerConstructor) {
-                                                        price={item.price}/>
-                                </div>
+                        {INGREDIENTS.map((item: BurgerItem | any, index: number) => (
+                            <li key={uuidv4()} className={styles.item}>
```
`key={uuidv4()`

в `key` нужно указывать только 1 раз созданный уникальный `uuid` ингредиента, иначе Вы сейчас меняете его при каждой перерисовке, что приводит к тому, что Реакт не может уследить за перемещением компонентов, и это может сломать логику `DnD`

А в `key` будет уже просто **key={item.uuid}**


--------------


Нужно генерировать uuid при создании экшена (желательно, в `action creator`). И созданный uuid должен приходить в редьюсер в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, так как проект не запускается. Скрин ошибки http://joxi.ru/DmBYyBDSq8j7zr  . Может, что-то не закоммитилось или не запушилось на гитхаб. Попробуйте скачать свой проект с гитхаба и запустить его

Еще заметил ошибки:

* чтобы экшен-криейтер создал экшен, нужно его вызывать
* в `key` нужно указывать только 1 раз созданный уникальный `uuid` ингредиента, иначе Вы сейчас меняете его при каждой перерисовке, что приводит к тому, что Реакт не может уследить за перемещением компонентов, и это может сломать логику `DnD`

maybe777 committed:
Поправлены замечания про использование UUID и вызов action creators

для фикса запуска проекта необходимо отредактировать webpack

/node_modules/react-scripts/config/webpack.config.js

alias: {
    "react/jsx-dev-runtime": "react/jsx-dev-runtime.js",
    "react/jsx-runtime": "react/jsx-runtime.js",
    }

gennady-bars reviewed (commented):
К сожалению, работу придется отклонить без потери итерации, так как проект не запускается. Скрин ошибки http://joxi.ru/Dr8YZBESKyjjJm

maybe777 committed:
Исправление ошибки запуска

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.



**Можно лучше**

* При кликах на табы `Булки, Соусы, Начинки` можно сделать логику `плавного` скролла до секции. Это можно сделать с помощью хука `useRef` и методом ` scrollIntoView`. Нужно установить для 3х секций 3 `useRef` и следить за изменением табов при кликах на них через функцию клика. Как меняется таб, то запускается ` scrollIntoView ` со значением ` behavior: "smooth"`
Подробнее про `useRef` тут https://ru.reactjs.org/docs/hooks-reference.html#useref
Подробнее про ` scrollIntoView` тут https://developer.mozilla.org/ru/docs/Web/API/Element/scrollIntoView
